// logless-client/lib/src/logless.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var logless_context_1 = require("./logless-context");
var Logless = (function () {
    function Logless() {
    }
    Logless.capture = function (source, handler) {
        if (handler === undefined || handler === null) {
            throw new Error("Handler is null or undefined! This must be passed.");
        }
        return new FunctionWrapper(source, handler).wrappingFunction();
    };
    Logless.logger = function (source) {
        return new logless_context_1.LoglessContext(source);
    };
    Logless.middleware = function (source) {
        var context = new logless_context_1.LoglessContext(source);
        if (Logless.captureConsole) {
            context.wrapConsole();
        }
        var capturePayloads = function (request, response, next) {
            context.newTransactionID();
            context.log(logless_context_1.LogType.INFO, request.body, null, ["request"]);
            Logless.wrapResponse(context, response);
            if (Logless.captureConsole) {
                context.captureConsole(function () {
                    next();
                });
            }
            else {
                next();
            }
        };
        var captureError = function (error, request, response, next) {
            context.logError(logless_context_1.LogType.ERROR, error, null);
            next();
        };
        capturePayloads.logger = context;
        captureError.logger = context;
        return new LoglessMiddleware(capturePayloads, captureError);
    };
    Logless.enableConsoleLogging = function () {
        Logless.captureConsole = true;
    };
    Logless.disableConsoleLogging = function () {
        Logless.captureConsole = false;
    };
    Logless.wrapResponse = function (context, response, onFlushed) {
        var originalEnd = response.end;
        response.end = function (data, encoding, callback) {
            var payload = data.toString();
            if (response.getHeader("content-type").startsWith("application/json")) {
                try {
                    payload = JSON.parse(payload);
                }
                catch (e) {
                    console.error("Could not parse JSON: " + payload);
                }
            }
            context.log(logless_context_1.LogType.INFO, payload, null, ["response"]);
            originalEnd.call(response, data, encoding, callback);
            context.flush();
        };
    };
    Logless.Domain = "logless.bespoken.tools";
    Logless.captureConsole = false;
    return Logless;
}());
exports.Logless = Logless;
var LoglessMiddleware = (function () {
    function LoglessMiddleware(requestHandler, errorHandler) {
        this.requestHandler = requestHandler;
        this.errorHandler = errorHandler;
    }
    return LoglessMiddleware;
}());
exports.LoglessMiddleware = LoglessMiddleware;
var FunctionWrapper = (function () {
    function FunctionWrapper(source, wrappedFunction) {
        this.source = source;
        this.wrappedFunction = wrappedFunction;
    }
    FunctionWrapper.prototype.handle = function (arg1, arg2, arg3) {
        if (arg2.awsRequestId !== undefined) {
            this.handleLambda(arg1, arg2, arg3);
        }
        else {
            this.handleCloudFunction(arg1, arg2);
        }
    };
    FunctionWrapper.prototype.handleLambda = function (event, context, callback) {
        var logger = new logless_context_1.LoglessContext(this.source);
        logger.onLambdaEvent(event, context, callback);
        context.logger = logger;
        try {
            this.wrappedFunction.call(this, event, context, logger.callback());
        }
        catch (e) {
            console.error(e);
            logger.flush();
            logger.cleanup();
        }
    };
    FunctionWrapper.prototype.handleCloudFunction = function (request, response) {
        var logger = new logless_context_1.LoglessContext(this.source);
        logger.onCloudFunctionEvent(request, response);
        request.logger = logger;
        try {
            this.wrappedFunction.call(this, request, response);
        }
        catch (e) {
            console.error(e);
            logger.flush();
            logger.cleanup();
        }
    };
    FunctionWrapper.prototype.wrappingFunction = function () {
        var wrapper = this.handle.bind(this);
        return wrapper;
    };
    return FunctionWrapper;
}());
