// bespoken-tools/lib/alexa/alexa-context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const alexa_session_1 = require("./alexa-session");
const uuid = require("uuid");
class AlexaContext {
    constructor(_skillURL, _interactionModel, _audioPlayer, _applicationID) {
        this._skillURL = _skillURL;
        this._interactionModel = _interactionModel;
        this._audioPlayer = _audioPlayer;
        this._applicationID = _applicationID;
        this._accessToken = null;
    }
    applicationID() {
        if (this._applicationID === undefined || this._applicationID === null) {
            this._applicationID = "amzn1.echo-sdk-ams.app." + uuid.v4();
        }
        return this._applicationID;
    }
    skillURL() {
        return this._skillURL;
    }
    interactionModel() {
        return this._interactionModel;
    }
    userID() {
        if (this._userID === undefined || this._userID === null) {
            this._userID = "amzn1.ask.account." + uuid.v4();
        }
        return this._userID;
    }
    setUserID(userID) {
        this._userID = userID;
    }
    accessToken() {
        return this._accessToken;
    }
    setAccessToken(token) {
        this._accessToken = token;
    }
    audioPlayer() {
        return this._audioPlayer;
    }
    audioPlayerEnabled() {
        return this._audioPlayer !== null;
    }
    newSession() {
        this._session = new alexa_session_1.AlexaSession();
    }
    session() {
        return this._session;
    }
    endSession() {
        this._session = null;
    }
    activeSession() {
        return this._session !== null;
    }
}
exports.AlexaContext = AlexaContext;
