// bespoken-tools/lib/alexa/interaction-model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const intent_schema_1 = require("./intent-schema");
const sample_utterances_1 = require("./sample-utterances");
class InteractionModel {
    constructor(intentSchema, sampleUtterances) {
        this.intentSchema = intentSchema;
        this.sampleUtterances = sampleUtterances;
    }
    static fromFiles(intentSchemaFile, sampleUtterancesFile, callback) {
        let callbackCount = 0;
        let callbackError = null;
        let intentSchema = null;
        let sampleUtterances = null;
        let done = function (schema, utterances, error) {
            callbackCount++;
            if (schema !== undefined && schema !== null) {
                intentSchema = schema;
            }
            if (utterances !== undefined && utterances !== null) {
                sampleUtterances = utterances;
            }
            if (error !== undefined && error !== null) {
                callbackError = error;
            }
            if (callbackCount === 2) {
                if (callbackError !== null) {
                    callback(null, callbackError);
                }
                else {
                    callback(new InteractionModel(intentSchema, sampleUtterances));
                }
            }
        };
        intent_schema_1.IntentSchema.fromFile(intentSchemaFile, function (schema, error) {
            if (error !== undefined && error !== null) {
                console.error("Error loading Intent Schema!");
                console.error("Cause: " + error);
                console.error();
            }
            done(schema, null, error);
        });
        sample_utterances_1.SampleUtterances.fromFile(sampleUtterancesFile, function (utterances, error) {
            if (error !== undefined && error !== null) {
                console.error("Error loading Sample Utterances!");
                console.error("Cause: " + error);
                console.error();
            }
            done(null, utterances, error);
        });
    }
    intentForUtterance(utterance) {
        return this.sampleUtterances.intentForUtterance(utterance);
    }
    hasIntent(intent) {
        return this.intentSchema.hasIntent(intent);
    }
}
exports.InteractionModel = InteractionModel;
