// bespoken-tools/lib/client/bst-virtual-alexa.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const global_1 = require("../core/global");
const virtual_alexa_1 = require("virtual-alexa");
const fs = require("fs");
class BSTVirtualAlexa {
    constructor(skillURL, interactionModel, intentSchemaFile, sampleUtterancesFile, applicationID, locale, userId) {
        this.skillURL = skillURL;
        this.interactionModel = interactionModel;
        this.intentSchemaFile = intentSchemaFile;
        this.sampleUtterancesFile = sampleUtterancesFile;
        this.applicationID = applicationID;
        this.locale = locale;
        this.userId = userId;
        this.virtualAlexa = null;
        this.interactionModelProvided = false;
        this.sampleUtterancesProvided = false;
        this.intentSchemaProvided = false;
        if ((intentSchemaFile || sampleUtterancesFile) && interactionModel) {
            console.error("The Interaction Model and Intent Schema Files should not both be specified. It should be one or the other.");
            throw new Error("The Interaction Model and Intent Schema Files should not both be specified. It should be one or the other.");
        }
        if (!this.interactionModel) {
            this.interactionModel = BSTVirtualAlexa.DefaultInteractionModelLocation;
        }
        else {
            this.interactionModelProvided = true;
        }
        if (!this.intentSchemaFile) {
            this.intentSchemaFile = BSTVirtualAlexa.DefaultIntentSchemaLocation;
        }
        else {
            this.intentSchemaProvided = true;
        }
        if (!this.sampleUtterancesFile) {
            this.sampleUtterancesFile = BSTVirtualAlexa.DefaultSampleUtterancesLocation;
        }
        else {
            this.sampleUtterancesProvided = true;
        }
        if (global_1.Global.config()) {
            if (this.applicationID) {
                global_1.Global.config().updateApplicationID(this.applicationID);
            }
            else {
                this.applicationID = global_1.Global.config().applicationID();
            }
        }
    }
    saveSession() {
        if (global_1.Global.config()) {
            const session = this.virtualAlexa.context().session();
            if (!session) {
                global_1.Global.config().deleteSession();
                return;
            }
            const savedSession = {
                id: session.id(),
                attributes: session.attributes(),
                locale: this.locale,
                userId: this.userId
            };
            global_1.Global.config().saveSession(savedSession);
        }
    }
    deleteSession() {
        global_1.Global.config().deleteSession();
    }
    loadSession() {
        if (global_1.Global.config()) {
            const savedSession = global_1.Global.config().loadSession();
            if (savedSession) {
                this.virtualAlexa.context().session().setID(savedSession.id);
                this.virtualAlexa.context().session().updateAttributes(savedSession.attributes);
            }
        }
    }
    context() {
        return this.virtualAlexa.context();
    }
    validateFile(file, fileType) {
        if (!fs.existsSync(file)) {
            console.error("Error loading " + fileType);
            console.error("Cause: '" + this.interactionModel + "' doesn't exist");
            throw new Error("Error loading " + fileType + ", file not found");
        }
    }
    validateJsonFiles(fileLocation, fileType) {
        const fileContent = fs.readFileSync(fileLocation);
        try {
            JSON.parse(fileContent.toString());
        }
        catch (error) {
            console.error("Error loading '" + fileType + "', incorrect JSON");
            console.error("Cause: ", error.message);
            throw new Error("Error loading '" + fileType + "', incorrect JSON");
        }
    }
    validateFilesAndBuild(createdEmptyInteractionModelIfNeeded) {
        const builder = virtual_alexa_1.VirtualAlexa.Builder().applicationID(this.applicationID).skillURL(this.skillURL);
        let usingInteractionModel = false;
        if (this.locale) {
            builder.locale(this.locale);
            if (!this.interactionModelProvided) {
                this.interactionModel = this.interactionModel.replace("en-US", this.locale);
            }
        }
        if (!(this.interactionModelProvided || this.intentSchemaProvided)) {
            if (fs.existsSync(this.interactionModel)) {
                usingInteractionModel = true;
            }
            else if (!(fs.existsSync(this.intentSchemaFile) && fs.existsSync(this.sampleUtterancesFile)) && !createdEmptyInteractionModelIfNeeded) {
                console.error("Error loading Interaction model, no file provided and none found in default locations");
                throw new Error("Error loading Interaction model, no file provided and none found in default locations");
            }
        }
        else {
            if (this.interactionModelProvided) {
                this.validateFile(this.interactionModel, BSTVirtualAlexa.FileTypes.InterationModel);
                usingInteractionModel = true;
            }
            else {
                if (!fs.existsSync(this.intentSchemaFile)) {
                    this.validateFile(this.intentSchemaFile, BSTVirtualAlexa.FileTypes.IntentSchema);
                }
                if (!fs.existsSync(this.sampleUtterancesFile)) {
                    this.validateFile(this.sampleUtterancesFile, BSTVirtualAlexa.FileTypes.SampleUtterances);
                }
            }
        }
        if (usingInteractionModel) {
            this.validateJsonFiles(this.interactionModel, BSTVirtualAlexa.FileTypes.InterationModel);
            builder.interactionModelFile(this.interactionModel);
        }
        else if (createdEmptyInteractionModelIfNeeded) {
            builder.interactionModel(BSTVirtualAlexa.DefaultInteractionModel);
        }
        else {
            this.validateJsonFiles(this.intentSchemaFile, BSTVirtualAlexa.FileTypes.IntentSchema);
            builder.intentSchemaFile(this.intentSchemaFile).sampleUtterancesFile(this.sampleUtterancesFile);
        }
        return builder.create();
    }
    start(createdEmptyInteractionModelIfNeeded) {
        this.virtualAlexa = this.validateFilesAndBuild(createdEmptyInteractionModelIfNeeded);
    }
    spoken(phrase, callback) {
        let request;
        this.virtualAlexa.filter((generatedRequest) => { request = generatedRequest; });
        this.loadSession();
        this.virtualAlexa.utter(phrase).then((payload) => {
            this.saveSession();
            if (callback !== undefined && callback !== null) {
                callback(null, payload, request);
            }
        }).catch(error => {
            callback(error, null, request);
        });
        return this;
    }
    intended(intentName, slots, callback) {
        let request;
        this.virtualAlexa.filter((generatedRequest) => { request = generatedRequest; });
        this.loadSession();
        this.virtualAlexa.intend(intentName, slots).then((payload) => {
            this.saveSession();
            if (callback !== undefined && callback !== null) {
                callback(null, payload, request);
            }
        }).catch(error => {
            callback(error, null, request);
        });
        return this;
    }
    launched(callback) {
        let request;
        this.virtualAlexa.filter((generatedRequest) => { request = generatedRequest; });
        this.virtualAlexa.launch().then(payload => {
            this.saveSession();
            callback(null, payload, request);
        }).catch(error => {
            callback(error, null, request);
        });
        return this;
    }
}
BSTVirtualAlexa.DefaultIntentSchemaLocation = "speechAssets/IntentSchema.json";
BSTVirtualAlexa.DefaultSampleUtterancesLocation = "speechAssets/SampleUtterances.txt";
BSTVirtualAlexa.DefaultInteractionModelLocation = "models/en-US.json";
BSTVirtualAlexa.DefaultInteractionModel = { interactionModel: { languageModel: { invocationName: "", intents: [] } } };
BSTVirtualAlexa.FileTypes = {
    InterationModel: "Interaction Model",
    IntentSchema: "Intent Schema",
    SampleUtterances: "Sample Utterances"
};
exports.BSTVirtualAlexa = BSTVirtualAlexa;
