// virtual-alexa/lib/src/SkillResponse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const _ = require("lodash");
class SkillResponse {
    constructor(rawJSON) {
        this.wrapJSON(rawJSON);
    }
    attr(key) {
        return _.get(this.sessionAttributes, key);
    }
    attrs(...keys) {
        return _.pick(this.sessionAttributes, keys);
    }
    card() {
        return _.get(this, "response.card");
    }
    cardContent() {
        return _.get(this, "response.card.content");
    }
    cardImage() {
        return _.get(this, "response.card.image");
    }
    cardSmallImageURL() {
        return _.get(this, "response.card.image.smallImageUrl");
    }
    cardLargeImageURL() {
        return _.get(this, "response.card.image.largeImageUrl");
    }
    cardTitle() {
        return _.get(this, "response.card.title");
    }
    promptSSML() {
        return _.get(this, "response.outputSpeech.ssml");
    }
    promptText() {
        return _.get(this, "response.outputSpeech.text");
    }
    repromptSSML() {
        return _.get(this, "response.reprompt.outputSpeech.ssml");
    }
    repromptText() {
        return _.get(this, "response.reprompt.outputSpeech.text");
    }
    wrapJSON(rawJSON) {
        for (const key of Object.keys(rawJSON)) {
            const value = rawJSON[key];
            this[key] = value;
        }
    }
}
exports.SkillResponse = SkillResponse;
