// bespoken-tools/lib/server/node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const socket_handler_1 = require("../core/socket-handler");
const logging_helper_1 = require("../core/logging-helper");
const Logger = "NODE";
class Node {
    constructor(id, socketHandler) {
        this.id = id;
        this.socketHandler = socketHandler;
        this.requests = {};
    }
    forward(request) {
        console.log("NODE " + this.id + " MSG-ID: " + request.id() + " Forwarding");
        this.requests[request.id()] = request;
        this.socketHandler.send(new socket_handler_1.SocketMessage(request.rawContents, request.id()));
    }
    handlingRequest() {
        return (Object.keys(this.requests).length > 0);
    }
    onReply(socketMessage) {
        const self = this;
        console.log("NODE " + this.id + " MSG-ID: " + socketMessage.getMessageID() + " ReplyReceived");
        const request = this.requests[socketMessage.getMessageID()];
        if (request === null) {
            logging_helper_1.LoggingHelper.info(Logger, "No matching messageID for reply: " + socketMessage.getMessageID());
        }
        else {
            delete self.requests[socketMessage.getMessageID()];
            try {
                request.sourceSocket.write(socketMessage.getMessage());
            }
            catch (e) {
                logging_helper_1.LoggingHelper.error(Logger, "Error writing: " + e);
            }
        }
    }
}
exports.Node = Node;
