// bespoken-tools/lib/server/webhook-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const webhook_request_1 = require("../core/webhook-request");
const http = require("http");
const https = require("https");
const logging_helper_1 = require("../core/logging-helper");
let Logger = "WEBHOOK";
class WebhookManager {
    constructor(port) {
        this.port = port;
        this.socketMap = {};
        this.onWebhookReceived = null;
        this.host = "0.0.0.0";
    }
    start(started) {
        let self = this;
        let socketIndex = 0;
        const connectFunction = function (socket) {
            let webhookRequest = new webhook_request_1.WebhookRequest(socket);
            socketIndex++;
            let socketKey = socketIndex;
            self.socketMap[socketIndex] = socket;
            socket.on("data", function (data) {
                webhookRequest.append(data);
                if (webhookRequest.done()) {
                    self.onWebhookReceived(webhookRequest);
                    webhookRequest = new webhook_request_1.WebhookRequest(socket);
                }
            });
            socket.on("close", function () {
                delete self.socketMap[socketKey];
            });
        };
        if (!process.env.SSL_CERT) {
            this.server = http.createServer().listen(this.port);
            this.server.on("connection", connectFunction);
        }
        else {
            const cert = process.env.SSL_CERT;
            const key = process.env.SSL_KEY;
            const credentials = {
                cert: cert.replace(/\\n/g, "\n"),
                key: key.replace(/\\n/g, "\n"),
            };
            const httpsServer = https.createServer(credentials);
            this.server = httpsServer.listen(this.port, this.host);
            this.server.on("secureConnection", connectFunction);
        }
        this.server.on("listening", function () {
            if (started !== undefined && started !== null) {
                started();
            }
        });
        logging_helper_1.LoggingHelper.info(Logger, "Listening on " + this.host + ":" + this.port);
    }
    stop(callback) {
        let self = this;
        for (let key in self.socketMap) {
            let socket = self.socketMap[key];
            socket.end();
        }
        this.server.close(function () {
            if (callback !== undefined && callback !== null) {
                callback();
            }
        });
    }
}
exports.WebhookManager = WebhookManager;
