// bespoken-tools/lib/external/source-name-generator.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const request_promise_native_1 = require("request-promise-native");
class SourceNameGenerator {
    callService() {
        return __awaiter(this, void 0, void 0, function* () {
            const options = {
                uri: "https://source-api.bespoken.tools/v1/sourceId",
                json: true,
                timeout: 30000
            };
            return request_promise_native_1.get(options);
        });
    }
    ;
    createDashboardSource(id, secretKey) {
        return __awaiter(this, void 0, void 0, function* () {
            const options = {
                uri: "https://source-api.bespoken.tools/v1/createSource",
                headers: {},
                body: {
                    source: {
                        id,
                        secretKey,
                        name: id,
                        liveDebug: true,
                    },
                },
                json: true,
                timeout: 30000
            };
            return request_promise_native_1.post(options);
        });
    }
}
exports.SourceNameGenerator = SourceNameGenerator;
